[TwigComponent] Ignore "nested" for Alpine & Vue attributes

smnandre opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Issues        | Fix #1839
| License       | MIT

Have lost time on Twig & the website :sweat:

Alternative implementation of #2325 

Update: Improved `__toString` performance following @Kocal comments

--- 
Now all these attributes are directly rendered 

| Framework       | Prefix | Code Example                                                                                       | Documentation                                                 |
|-----------------|---------|-------------------------------------------------------------------------------------------------------|----------------------------------------------------------------|
| **Alpine.js**   | `x-`    | `<div x-data="{ open: false }" x-show="open"></div>`                                                 | [Documentation Alpine.js](https://alpinejs.dev/)               |
| **Vue.js**      | `v-`    | `<input v-model="message" v-if="show">`                                                              | [Documentation Vue.js](https://vuejs.org/guide/)               |
| **Stencil**     | `@`     | `<my-component @onClick="handleClick"></my-component>`                                               | [Documentation Stencil](https://stenciljs.com/docs/)           |
| **Lit**         | `@`     | `<button @click="${this.handleClick}">Click me</button>`                                             | [Documentation Lit](https://lit.dev/docs/)                    |

smnandre committed:
[TwigComponent] Ignore "nested" for Alpine & Vue attributes

Have lost time on Twig & the website :sweat:

smnandre committed:
fabbot

Kocal commented on src/TwigComponent/src/ComponentAttributes.php:
```diff
@@ -41,7 +43,11 @@ public function __toString(): string
             ),
             function (string $carry, string $key) {
-                if (preg_match(self::NESTED_REGEX, $key)) {
+                if (
+                    preg_match(self::NESTED_REGEX, $key)
+                    && !preg_match(self::ALPINE_REGEX, $key)
+                    && !preg_match(self::VUE_REGEX, $key)
+                ) {
                     return $carry;
                 }
```
`preg_match` calls can be expensive, and can be avoid by quickly checking if the attribute name contains a `:` (ex: https://github.com/symfony/symfony/pull/28106). If it's not the case we early exit (without calling `preg_match`), otherwise we continue.

```suggestion
                if (
                    str_contains($key, ':')
                    && preg_match(self::NESTED_REGEX, $key)
                    && !preg_match(self::ALPINE_REGEX, $key)
                    && !preg_match(self::VUE_REGEX, $key)
                ) {
                    return $carry;
                }
```

Can you add the same check to `\Symfony\UX\TwigComponent\ComponentAttributes::nested` please? 

Thanks!

PS: Not related to your PR, but I'm not a fan about all the `array_reduce`/`array_filter`/`array_keys` calls where a `foreach` could be enough. We can open a new PR to rewrite and optimize that :)

smnandre replied:
> Can you add the same check to \Symfony\UX\TwigComponent\ComponentAttributes::nested please?

The str_contains will be enough here, as if someone want to explicitely ask for .nested("x-on") let's not block this.

> I'm not a fan about all the array_reduce/array_filter/array_keys calls where a foreach could be enough. We can open a new PR to rewrite and optimize that :)

Yes, will do leter

smnandre replied:
Done here in the end, i'll let you check

smnandre committed:
Performance

Co-authored-by: Hugo Alliaume <hugo@alliau.me>

smnandre committed:
Add str_contains in nested method

smnandre committed:
Improve __toString performance

Running scenario: Short attributes, empty rendered
  __toString() time: 0.00077986717224121 seconds
  __toString2() time: 0.00049495697021484 seconds
  Improvement: 36.533170284317%
Running scenario: Short attributes, partial rendered
  __toString() time: 0.00054383277893066 seconds
  __toString2() time: 0.00031709671020508 seconds
  Improvement: 41.692240245506%
Running scenario: Short attributes, full rendered
  __toString() time: 0.00030899047851562 seconds
  __toString2() time: 0.00012898445129395 seconds
  Improvement: 58.256172839506%
Running scenario: Long attributes, empty rendered
  __toString() time: 0.0038020610809326 seconds
  __toString2() time: 0.0026099681854248 seconds
  Improvement: 31.353859660124%
Running scenario: Long attributes, partial rendered
  __toString() time: 0.0032980442047119 seconds
  __toString2() time: 0.0022611618041992 seconds
  Improvement: 31.439311790646%
Running scenario: Long attributes, full rendered
  __toString() time: 0.00074219703674316 seconds
  __toString2() time: 0.00014185905456543 seconds
  Improvement: 80.886604561516%

WebMamba commented:
I am sorry but I don't like the solution taken here. I like way more the initial solution taken here: #2325. I think we should not add code specific to another framework, and I don't think that we should ignore nested attributes without making them implicit for the user. I think we should focus on a solution like the initial solution where the user choose to ignore it implicitly.

Kocal commented:
What about https://github.com/symfony/ux/pull/2325#issuecomment-2453408427 then?
People will have to adapt their code to make it works with Twig Components?

I don't find that super friendly, it not obvious at all, and can be a source of frustration to the user even if it's documented (and document that you need to explicitly use a new syntax to disable nested attributes to makes your code works, feels like a loss to me)

smnandre commented:
Are we _really_ considering there is real scenario where someone 
* has a component
* has a child component in it named "x-on" 
* uses nested attributes

?

Regarding specific code, we have some for LiveComponent & Alpine already. On the other hand, opening a new char / syntax in the attributes will be a change with many more problems.

Kocal commented:
Thank you @smnandre for taking care of this regression.
